### ISSUE TITLE
Extend Build Systems guide with background information 

### ISSUE THREAD

**Issue by omsai** [LLNL/spack#23123]
TL;DR: Is there interest in teaching new spack contributors more generally about build systems, and can we bring that into spack's contributor documentation?

One of the most fun experiences I've had with packaging software for Gentoo Linux was reading through their comprehensive [developer manual](https://devmanual.gentoo.org/) which extends the packaging mission of the Gentoo project to more generally teach you critical concepts about build systems.  For example they have a section that explains [GNU Autotools](https://devmanual.gentoo.org/general-concepts/autotools/index.html) which is one of the most succinct and useful references to remind oneself of how Autotools works and to reason about patching and troubleshooting errors.  On that page, you can see their rationale is:

>  Finding and correcting build issues is often easier if the build system is not seen simply as a scary black box. 

Currently, contributors to spack are slightly abstracted away from the build system, and we need to lower the barrier both of where to look for log files, how to make and test changes inside and outside of the `package.py`, and generally lower the expertise in the build system they are working on.

1. Contributors may not have written the original build system of the project; turnover of developers is not unusual for academic software, and
2. Contributors may primarily be researchers with a little bit of experience compiling software and may be motivated to use spack's dependencies and know python but lack expertise with build systems.

I know when I started out with packaging software for Gentoo I was both learning the build system and the package manager at the same time.  It, of course, helps to read similar packaged software, but there is a gap to fill between the official build system documentation and reading similar work.

I feel like this might be a fun thing to do as a group at a hackathon / codeathon.

**Comment by alalazo** [comment]
Another useful resource for Autotools is [here](https://autotools.io/)

**Comment by adamjstewart** [comment]
I wrote basically all of our current [Build Systems](https://spack.readthedocs.io/en/latest/build_systems.html) docs, so here are my thoughts.

### Cons

I'm a bit afraid to include this in Spack's documentation because it isn't really specific to Spack. The current build system docs explain a bit about each build system, but only enough to write a Spack package for it. Our documentation is already so long that most users don't read _all_ of it. It also becomes harder to maintain the longer the docs get.

### Pros

It really isn't possible to write a package for [MakefilePackage](https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html) without some basic understanding of how Makefiles work. You almost always have to modify the Makefile in some way (to use Spack's compiler wrappers or Spack-installed dependencies), so your proposal makes quite a lot of sense for [MakefilePackage](https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html) at least.

I already wrote quite a lot about the history of Python packaging in our [PythonPackage](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html) docs, as it's needed just to understand where to find a list of dependencies.

### Alternatives

More extensive documentation on each build system could be added as external links. For example, I'm happy to add links like https://autotools.io/ to https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html#external-documentation.

Overall, I'm fine with adding anything that helps people write a Spack package, and adding additional external links as needed.

**Comment by omsai** [comment]
@alalazo That's one of my favorite autotools resources!  I actually ran into the author on one of Adafruit's Discord channels last year and thanked him for the excellent resource; his day job  has since taken him away from that type of work and he lamented not completing the online book as much as he originally intended.  It's been adequate for my purposes though.

@adamjstewart Yes, looking at it now I'm more wary of the high ratio of the amount of work involved and added maintenance burden to the intended pay off - so perhaps a simpler instructive type of documentation might be to add 1 or, at most, 2 full tutorial examples in those build system pages of taking a project tarball with a low-to-moderate amount of complexity and writing the spack package.  Basically, creating the `package.py` and working through adding the upstream dependencies, reading the configure and build output files, talking about error messages and poking around the temporary build directory.  One would probably have to pin to a particular version of a package.

What I'm not suggesting is writing this for all 18 `*Package`-class build systems at the get-go.  We could each write an example for one build system, like say a MakefilePackage, CMakePackage, or PythonPackage, etc.  For less common build systems we could even ask other spack package authors what they think are good packages that compile quickly enough to be helpful as an example.  I feel we should strive to use authentic examples of real world packages: in the past, I've written tutorials for packaging small toy programs which might be end being a fall back.

At the risk of extending the package pages and creating git noise, we could probably work off of a separate project like "Spack packages by example" and either merge it into the docs or if they're too long, maybe have a standalone external repository like the spack tutorial currently is.

If this seems worthwhile, we would probably want to impose limits on how long running the a tutorial example should take: maybe 2 hours?  The next steps would be to write out the learning objectives, some of the package manager concepts we would like to touch on, and elements of the workflow we would like to include. Then shortlist some nice existing packages, or other packages not yet packaged in spack where packaging them is nevertheless instructive.
